ai_spider/stats.py

```diff
@@ -6,30 +6,176 @@
 from threading import Thread
 
 from ai_spider.unique_queue import UniqueQueue
+from sqlalchemy import and_
 
 STATS_EMA_ALPHA = 0.9
 PUNISH_SECS = 60 * 15
 PUNISH_BAD_PERF = 9999
 
+class StatsStore:
+    def __init__(self, session=None):
+        self.update_queue = UniqueQueue(key=lambda el: el[0])
+        self.update_thread = Thread(daemon=True, target=self.update_loop)
+        self.update_thread.start()
+        self.session=session
+
+    def stop(self, join=True):
+        self.update_queue.put([None, None])
+        if join:
+            self.update_thread.join()
+
+    def update_loop(self):
+        while True:
+            try:
+                key, bnum, val = self.update_queue.get(timeout=4)
+                if key is None:
+                    break
+                self._update(key, bnum, val)
+            except Empty:
+                pass
+
+    def update(self, key: str, bin_num: int,  val: dict):
+        self.update_queue.put([key, bin_num, val])
+
+    def _update(self, key: str, bin_num: int, vals: dict):
+        raise NotImplementedError
+
+    def get(self, key: str, bin_num: int, filters: dict) -> dict:
+        raise NotImplementedError
+    
+    def create(self, vals: dict):
+        raise NotImplementedError
+class MySqlWorkerStats(StatsStore):
+    def __init__(self, connection=None):
+        self.connection=connection
+        
+    def create(self, vals: dict):
+        insert_query=("""
+        INSERT INTO workers (stats_container_id, msize, val)
+        VALUES (%s, %s, %s)
+        RETURNING *
+        """)
+        cursor = self.connection.cursor()
+        cursor.execute(insert_query, (vals["row_id"], vals["msize"], vals["val"]))
+        self.connection.commit()
+        return cursor.fetchone()
+    
+    def get(self, key: str, bin_num: int, filters: dict) -> dict:
+        get_query=("""
+            SELECT * FROM worker_stats where stats_container_id = %s AND msize = %s
+        """)
+        cursor = self.connection.cursor()
+        cursor.execute(get_query, (bin_num, filters["msize"]))
+        self.connection.commit()
+        return cursor.fetchall()
+    
+    def _update(self, key: str, bin_num: int, vals: dict):
+        update_query = ("""
+            UPDATE worker_stats
+            SET val = %s
+            WHERE id = %s
+        """)
+        cursor = self.connection.cursor()
+        cursor.execute(update_query, (vals["secs"],vals["row_id"]))
+        self.connection.commit()
+    
+    def bump_bin(self, row_id: int, secs: int):
+        dict = {
+            "row_id": row_id,
+            "secs": secs
+        }
+        self.update(val = dict)
+        
 
 class StatsBin:
-    def __init__(self, alpha, val=None):
+    def __init__(self, alpha, worker_id: int, msize: int, val=None, store: MySqlWorkerStats = None):
+        self.msize=msize
+        unique_fields = {
```

Comment:
remove this.  store should not be passed in.

Reply:
how will the get function filter down which rows to get with out it?